integration-tests/slack/slack-utils.js

```diff
@@ -0,0 +1,139 @@
+  } else if (!token) {
+    throw new Error(`**  ERROR: Missing CircleCI API token. Failed to fetch CircleCI job artifacts.`)
+  }
+
+  // Find more arguments and details of the response: https://circleci.com/docs/api/v2/index.html#operation/getJobArtifacts
+  const apiUrl = 'https://circleci.com/api/v2/project/github/DataBiosphere/terra-ui'
+  const headers = { Accept: 'application/json', Authorization: `Bearer ${token}` }
+
+  try {
+    const response = await fetch(`${apiUrl}/${buildNum}/artifacts`, { method: 'GET', headers })
+    const json = await response.json()
+    const { items } = json
+    const filtered = _.filter(item => item['path'].toString().includes('tests-summary'), items)
+    return filtered.map(item => item.url)
+  } catch (e) {
+    console.error(`**  ERROR: Encountered unexpected error when getting CircleCI build_number: ${buildNum} artifacts:`, e)
+    throw e
+  }
+}
+
+const getFailedTestFileName = aggregatedResult => {
+  const failedTests = []
+  _.forEach(testResult => {
+    if (testResult.numFailingTests > 0) {
+      failedTests.push(parse(testResult.testFilePath).name)
+    }
+  }, aggregatedResult.testResults)
+  return failedTests
+}
+
+const getFailedTests = async () => {
+  const headers = { Accept: 'application/json' }
+  const urls = await fetchCircleJobArtifacts()
+  const tests = []
+  for (const url of urls) {
+    try {
+      // Parse all tests-summary JSON for failed tests
+      const response = await fetch(url, { method: 'GET', headers })
+      const json = await response.json()
+      const failedTest = getFailedTestFileName(json)
+      tests.push(...failedTest)
+    } catch (e) {
+      console.error(`**  ERROR: Encountered unexpected error when getting CircleCI artifacts tests-summary file: ${url}.`, e)
+      throw e
+    }
+  }
+  return tests
+}
+
+const getChannelsNotifyFailed = failedTests => {
+  const channels = new Map()
+  const data = JSON.parse(fs.readFileSync('./slack/slack-notify-channels.json', 'utf8'))
+  const { fail: jsonItems } = data
+
+  _.forEach(item => {
+    const id = _.get(['id'], item)
+    const { tests } = item
+    const names = _.map(item => item.name, tests)
+    channels.set(id, names)
+  }, jsonItems)
+
+  // Construct a map object: key: channel_id, value: test_names array
+  const idsMap = new Map()
+  _.forEach(test => {
+    const filtered = Array.from(channels.keys())
+      .map(key => channels.get(key).includes(test) || channels.get(key).length === 0 ? key : undefined)
+      .filter(key => key)
+    if (filtered.length === 0) {
+      throw new Error(`Test: ${test} was not found in slack-notify-channels.json`)
+    }
+    _.forEach(k => idsMap.has(k) ? idsMap.get(k).push(test) : idsMap.set(k, new Array(test)), filtered)
+  }, failedTests)
+
+  return idsMap
+}
+
+/**
+ * Notify job failed. Message contains failed test names.
+ * @param failedTests
+ * @returns {Promise<void>}
+ */
+// eslint-disable-next-line
```

Comment:
eslint warning. I don't know how to resolve it.
```
104:41  warning  Async arrow function has no 'await' expression  require-await
```

Reply:
But `await postMessage` is called inside the `_.forEach`. I'll try remove it to see if still works.

Reply:
Yes, but this function is not awaiting the functions called by `_.forEach`.